server/src/main/java/org/elasticsearch/action/ingest/WriteableIngestDocument.java

```diff
@@ -67,19 +134,25 @@ IngestDocument getIngestDocument() {
 
     @Override
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
-        builder.startObject("doc");
-        Map<IngestDocument.MetaData, Object> metadataMap = ingestDocument.extractMetadata();
+        builder.startObject(DOC_FIELD);
+        Map<IngestDocument.MetaData, Object> metadataMap = ingestDocument.getMetadata();
         for (Map.Entry<IngestDocument.MetaData, Object> metadata : metadataMap.entrySet()) {
             if (metadata.getValue() != null) {
                 builder.field(metadata.getKey().getFieldName(), metadata.getValue().toString());
             }
         }
-        builder.field("_source", ingestDocument.getSourceAndMetadata());
-        builder.field("_ingest", ingestDocument.getIngestMetadata());
+        Map<String, Object> source = IngestDocument.deepCopyMap(ingestDocument.getSourceAndMetadata());
+        metadataMap.keySet().forEach(mD -> source.remove(mD.getFieldName()));
+        builder.field(SOURCE_FIELD, source);
+        builder.field(INGEST_FIELD, ingestDocument.getIngestMetadata());
```

Comment:
question: are you addressing collateral bugs that you found here? I mean the deepCopyMap and the extractMetadata that mutates the map.

Reply:
Kind of. If you call `toXContent` on the same object twice you will get different results because the extractMetadata method loses the metadata info. I do not like this. Also for the tests, this is bad because they call `toXContent` and then parse the object using `fromXContent` and compare the two objects. But the initial one doesn't contain the metadata anymore and the equality fails.

Reply:
gotcha, thanks for explaining, this seems pretty bad indeed.

Reply:
I think a better solution would be to change the `IngestDocument` class. We could extract the metadata into another map and then use that. But I didn't know how existing code interacted with it so I left it alone.

Reply:
I think I like isolating this change to the parsing because it is big enough as it stands. But I think it'd be worth looking at `IngestDocument` in a followup.